### PR TITLE
6139 FHIRtoHL7  Duplicate Element Name Protection

### PR DESCRIPTION
**Pull request by TurkeyFried**
This PR prevents multiple elements in a schema from having the same name.

Test Steps:
1. Run the converter with the `fhirdata` command. e.g. `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir -s metadata/hl7_mapping/ORU_R01/ORU_R01-base.yml --output-format=HL7`.
2. Verify the output is valid HL7.
3. Open any FHIRtoHL7 config (i.e. `patient-visit.yml`) and change the `name` value on multiple elements to be the same.
4. Repeat step 1, but now an error will appear pointing at the culprit.

## Changes
- Added a safety check for duplicate element names with an exception that shows the troublemakers.
- Added a new test case to cover this.

## Linked Issues
- #6139

### CODE REVIEW

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt [hunk 1]
```diff
@@ -98,6 +98,14 @@ class FhirToHl7Converter(
         // Add any schema level constants to the context
         // We need to create a new context, so constants exist only within their specific schema tree
         val schemaContext = CustomContext.addConstants(schema.constants, context)
+
+        // check for duplicate named schema elements before processing
+        val dupes =
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt:103]: You want to do this check while validating the schema and not when processing data.

File: prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirToHl7ConverterTests.kt [hunk 2]
```diff
@@ -363,5 +363,26 @@ class FhirToHl7ConverterTests {
...
+
+        val dupe = ConfigSchemaElement(
+            "iMustBeUnique",
+            value = listOf(pathWithValue),
+            hl7Spec = listOf("MSH-12")
+        )
+        schema = ConfigSchema(
+            hl7Type = "ORU_R01",
+            hl7Version = "2.5.1",
+            elements = mutableListOf(element, dupe)
```
Comment (carlosfelix2) [inline prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirToHl7ConverterTests.kt:382]: Add a test where you check for duplicate elements in a schema containing a schema.  Also add another to make sure the same element name in one leg of a schema tree is not a duplicate in another leg.

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]

File: prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirToHl7ConverterTests.kt (hunk 2 above)
Comment (TurkeyFried) [reply prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirToHl7ConverterTests.kt:382]: Added!

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt (hunk 1 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt]: Moved the check up the chain

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt [hunk 3]
```diff
@@ -78,6 +78,14 @@ class FhirToHl7Converter(
         // Sanity check, but at this point we know we have a good schema
         check(message != null)
         terser = Terser(message)
+        val dupes = (
+            schemaRef.elements.filter { it.name != null } +
+                schemaRef.elements.flatMap { it.schemaRef?.elements ?: emptyList() }
+            ).groupingBy { it.name }.eachCount().filter { it.value > 1 }
+
+        if (dupes.isNotEmpty()) { // value is the number of matches
+            throw SchemaException("Schema ${schemaRef.name} had multiple elements with the same name: $dupes")
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt:87]: Change to: Schema x HAS ...

Comment (carlosfelix2) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt:87]: The $dupes portion shows this: Schema ORU_R01-base had multiple elements with the same name: {set-id=2}
What is set-id=2?  Can we cleanup that last portion to make it more readable?  Not sure what the 2 is.

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt (hunk 1 above)
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt]: I do not see the move.  It is still done in the convert.  SHould be done as part of the schema validate.

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]
Tested and works great.

File: prime-router/metadata/hl7_mapping/ORU_R01/base/datatype/xon-organization.yml [hunk 4]
```diff
@@ -3,19 +3,19 @@ elements:
     value: [ '%resource.name' ]
     hl7Spec: [ '%{hl7OrgField}-1' ]
 
-  - name: xon-organization-type
+  - name: xon-organization-name-type
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/datatype/xon-organization.yml:6]: Looks like there is one more dupe.  Look for set-id in patient.  Thanks for cleaning all of these.

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt (hunk 3 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt]: Updated it to be clearer

File: prime-router/metadata/hl7_mapping/ORU_R01/base/datatype/xon-organization.yml (hunk 4 above)
Comment (TurkeyFried) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/datatype/xon-organization.yml:6]: Fixed all missing dupes

**Review by carlosfelix2 — APPROVED** [review]

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt [hunk 5]
```diff
@@ -98,6 +106,7 @@ class FhirToHl7Converter(
         // Add any schema level constants to the context
         // We need to create a new context, so constants exist only within their specific schema tree
         val schemaContext = CustomContext.addConstants(schema.constants, context)
+
```
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt:109]: Is there some kind of standard for these kinds of formatting changes on the team? Have we discussed using a common formatter/linter?

Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt:109]: We're using klinter with some standard settings that are stored....somewhere. On Intellij you should get lots of notices after installing the plugins
